### PR TITLE
Granular configuration of auto-endorsement 

### LINKED ISSUE #32: Granular endorser configuration
The endorser service supports auto-endorsing specific types of transaction, however it is not designed to support more fine-grained control mechanisms that would be necessary when running in production.

This issue aims to outline the desired target behaviour, and propose a design that should accommodate most if not all all configuration needs.

## Assumptions
* Endorsers can be configured globally to automatically accept incoming connections and to auto-endorse transactions using settings exposed by ACA-Py (see [here](https://github.com/hyperledger/aries-cloudagent-python/blob/73ba85d07fd1b557d6f6cf4d3701e733bf3d3e3f/Endorser.md))
* Endorser can automatically approve specific types of transactions (see https://github.com/hyperledger/aries-endorser-service/pull/23))
* Endorsers for production networks should follow a "deny by default" policy: new connection requests and transaction endorsement requests will always require approval from an endorser administrator.

## Improvements
* Connections originating from known DIDs or hosts (e.g.: in case of tenants of a known agency) may be automatically accepted following an allow list.
* A permission matrix will be used to define whether the DID can be published to the ledger or not, and how to handle endorsement requests for Schema/CredDef/Revreg/RevRegEntry (`AUTO`, `MANUAL`, `DENY`)
* An allow list will be used to further narrow down which schemas a specific DID is allowed to request endorsement for. Wildcards (`*`) can be used to make the restriction looser.
* An allow list will be used to further narrow down which credential definitions a specific DID is allowed to request endorsement for. Wildcards (`*`) can be used to make the restriction looser.

_Note:_ I can't think of a good reason to prevent an issuer from publishing a credential definition (and related revocation registry and entries), so the assumption is made that revocation-related endorsement requests will follow the rules and permissions enforced by the ledger, without additional filtering/handling. More thinking may be required if there are good reasons to selectively block/allow endorsement requests for these object types.

### Data structures
The following is an initial draft of the data structures required to support the above functionality.

**connection-allow-list**
|Source|
|---|
|https://my-multi-tenant.something.domain |
|XpgeQa93eZvGSZBZef3PHn|

**permission-matrix**
|DID|Public|Schema|CredDef|
|---|---|---|---|
|Ui6HA36FvN83cEtmYYHxrn|True|MANUAL|AUTO
|NCwGwDrzbZEqesYQummzWW|False|MANUAL|MANUAL|

**schema-allow-list**
|DID|Name|Version|
|---|---|---|
|Ui6HA36FvN83cEtmYYHxrn|my-schema|0.1.0|
|Ui6HA36FvN83cEtmYYHxrn|my-schema|0.2.0|
|NCwGwDrzbZEqesYQummzWW|another-schema|*|
|NCwGwDrzbZEqesYQummzWW|my-schema|0.2.0|

**credential-definition-allow-list**
|DID|Schema Issuer DID|Schema Name|Schema Version|Tag|
|---|---|---|---|---|
|Ui6HA36FvN83cEtmYYHxrn|Ui6HA36FvN83cEtmYYHxrn|my-schema|0.2.0|demo|
|NCwGwDrzbZEqesYQummzWW|Ui6HA36FvN83cEtmYYHxrn|my-schema|0.2.0|something-else|

### Updating and managing state
The endorser will expose secured API endpoints to update each one of the data structures. The data will be uploaded to the service as a `.csv` file (with header row) containing the latest snapshot data for the permissions and allow lists. The update process will drop the previous table contents and load the new data. Mechanisms to prevent/limit failure due to partially bad data should be implemented.

This approach has the the following benefits:
* it allows permission data to be managed externally, without the (immediate) requirement for a UI
* it allows bulk data loading and updating via ETL-like processes
* a UI could be easily implemented by extending the endpoints to support basic CRUD operations
* it allows for the source data to be kept private, if necessary, and pushed to the endorser as-needed

*Example:* the `.csv` files are maintained and stored in a GitHub repository. Pull requests are used to update them, allowing an easy review process for changes. GitHub natively displays `.csv` files as tables, allowing searching and filtering. On `merge`, a GHA can be triggered to automatically POST the new files to the target endorser(s).

### PR DESCRIPTION
**Pull request by Gavinok**
This pull request resolves #32

The endorser service supports auto-endorsing specific types of transaction, however it is not designed to support more fine-grained control mechanisms.

Here we add a new `/allow/` endpoint which can add to, delete from, and get a list of transactions which will be auto-endorsed. There is a list for:
- publishing DIDs
- publishing schema definitions
- publishing credential definitions

When any of these lists are updated all pending transactions will be compared and if a match between one of these lists and the pending transactions is found the transaction will be endorsed.

In addition, a new configuration setting `ENDORSER_REJECT_BY_DEFAULT` has been introduced to control whether transactions that do not match any of the allow lists are rejected, or left pending by default

Currently remaining work to be done:

### CODE REVIEW

**Review by esune — CHANGES_REQUESTED** [review]
Just a couple of small comments/tweaks, looking great otherwise!

File: endorser/api/db/models/allow.py [hunk 1]
```diff
@@ -108,14 +107,15 @@ class AllowedCredentialDefinition(BaseModel, table=True):
         )
     )
 
+    # acapy data ---
     issuer_did: str = Field(nullable=False, default=None)
     author_did: str = Field(nullable=False, default=None)
     schema_name: str = Field(nullable=False, default=None)
     version: str = Field(nullable=False, default=None)
     tag: str = Field(nullable=False, default=None)
     # TODO change to boolean
```
Comment (esune) [inline endorser/api/db/models/allow.py:116]: Remove the TODO now? 🙂

**Review by esune — CHANGES_REQUESTED** [review]
Final tweak, remaining changes look good to me

File: endorser/api/db/models/allow.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+Models of the Endorser tables for allow list and related data.
+
+"""
+import uuid
+from datetime import datetime
+
+from sqlmodel import Field
+from sqlalchemy import Column, func, text
+from sqlalchemy.dialects.postgresql import UUID, TIMESTAMP
```
[the remark this answers is not in the document]
Comment (esune) [reply endorser/api/db/models/allow.py:11]: If it is broadly used in the project we may as well leave it, I didn't realize it was already being used elsewhere and thought it was a new addition.

File: endorser/api/db/models/allow.py (hunk 1 above)
Comment (Gavinok) [reply endorser/api/db/models/allow.py:116]: How did I miss this 💀

**Review by esune — APPROVED** [review]
Looks good to me! 👍🏻

**Review by swcurran — APPROVED** [review]
Approved based on @esune ’s review.

**Pull request merged**